fe/reweighting.py

```diff
@@ -0,0 +1,138 @@
+
+from jax import vmap, numpy as np
+from jax.scipy.special import logsumexp
+
+from typing import List
+
+
+def ESS(log_weights: np.array) -> float:
+    """Effective sample size available for reweighting using a collection of log importance weights.
+
+    Notes
+    -----
+    * This uses the conventional definition ESS(w) = 1 / \sum_i w_i^2, which has some important known limitations!
+    * See [Elvira, Martino, Robert, 2018] "Rethinking the effective sample size" https://arxiv.org/abs/1809.04129
+        and references therein for some insightful discussion of limitations and possible improvements
+    """
+    log_Z = logsumexp(log_weights)
+    weights = np.exp(log_weights - log_Z)  # sum(weights) == 1
+    return 1 / np.sum(weights ** 2)  # between 1 and len(weights)
+
+
+class ReweightingLayer:
+    def __init__(self, x_k: List[np.array], u_fxn: callable, ref_params: np.array, lambdas: np.array):
+        """Assumes samples x_k[k] are drawn from e^{-u_fxn(x, lambdas[k], ref_params)}.
+
+        The constructor will
+        * aggregate all the samples into a flat structure xs (dropping information about which lambda window each sample
+            came from),
+        * precompute a matrix "u_kn" containing u_fxn(x, lam, ref_params) for x in xs, lam in lambdas, and
+        * pass this matrix to pymbar.MBAR to form self-consistent estimates for the free energies of all lambda windows.
+
+        These free energy estimates in turn imply statistical weights of each sample x in a weighted mixture of the
+        sampled lambda windows. These "importance weights" allow us to re-use the samples xs to compute arbitrary
+        expectations in thermodynamic states that have sufficient "overlap" with this weighted mixture.
+
+        Crucially, if we know how to differentiate U_fxn w.r.t. params, then
+        we also know how to differentiate the importance weights w.r.t. params.
+
+        In particular, to compute a differentiable estimate for the free energy difference between lam=1.0 and lam=0.0,
+        using the same potential energy function u_fxn but a possibly new parameter set params, we need to be able to
+        compute u_fxn(x, lam=0, params), u_fxn(x, lam=1, params) and their gradients w.r.t. params on all cached xs.
+
+        References
+        ----------
+        * Shirts MR and Chodera JD. Statistically optimal analysis of samples from multiple equilibrium states.
+            J. Chem. Phys. 129:124105, 2008. http://dx.doi.org/10.1063/1.2978177
+        * Shirts MR. Reweighting from the mixture distribution as a better way to describe the Multistate Bennett Acceptance Ratio.
+            arXiv preprint, 2017. https://arxiv.org/abs/1704.00891
+        * pymbar implementation of computePerturbedFreeEnergies , which has an associated uncertainty estimate as well!
+            https://github.com/choderalab/pymbar/blob/3c4262c490261110a7595eec37df3e2b8caeab37/pymbar/old_mbar.py#L1106-L1212
```

Comment:
We already discussed this, just a reminder to point to the `mbar.py` implementation rather than `old_mbar.py`

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/404/commits/8b81efe1d8e92122cb1cda2107bd58930c5fadd5